### ISSUE TITLE
[NPM] multiple network policies causing azure-npm to fill node disks by excessive logs 

### ISSUE THREAD

**Issue by bergerx** [Azure/azure-container-networking#598]
**What happened:**  

TL;DR: azure-npm logs directly into the node's root fs, it logs too many lines per rule, any no log rotation applied to the log file. Eventually that single log file fills the underlying node's root disk and triggers pod evacuations. AKS keeps replacing nodes as they are not able to handle issues to some point but eventually you'll find majority of the nodes under flaky DiskPressure and most workload fails to be scheduled and stable, rendering cluster to work in degraded mode.

More detailed explanation:

We deployed per-namespace Network Policies (4 per namespace, ~80 namespaces, ~330 in total) and they in-turn triggered lots of iptables rules to be generated. Azure support told us its too many for a cluster and causing thousands of iptables rules to be generated in total. There were actually ~3k iptables rules on each node (and at a later stage removing all network policies reduced that number to ~2k). We tried merging them into one but that didn't cause that dramatic change on the total amount of iptables rules. While the 4 NetworkPolicies were causing 12 iptables rules to be added, the merged version of that 4 was still causing 7 rules, this means current ~3k iptables rule count could be reduced to ~2.5k by applying the merged versions. We also don’t really know about what’s the real impact of having ~3k iptables rules by itself, since having several thousand iptables rules doesn’t seem brutal (of course depends of the rules but anyway). 

But apparently the indirect impact of having that many rules is problematic due to these two issues on azure-npm:
* azure-npm logs all actions taken for each rule, which slowly degrades the underlying node and eventually renders a fatal issue over time.
* azure-npm is seemingly keeping all rules in memory during reconciliation of the iptables rules (not sure), and the azure-npm pods were getting OOMKilled due to the fact, this is minor since it eventually catches up,
* azure-npm pods start to consume lots of cpu and started to be cpu throttled aggressively, again seems like a minor issue since it can catch up over time

The main major impact of having that many NetworkPolicies is triggered by the excessive logs written by azure-npm into the node’s root file system (/var/log/azure-npm.log). Which eventually caused all nodes' root disks to get full, once the nodes are getting the root disks full, the DiskPressure condition of nodes goes off and triggers several internal mechanisms including image GC and after then pod evacuations. The kubelet's actions for mitigating the DiskPressure issue cant help with the status and just defers the unavoidable result and also causing pods to keep jumping around due to triggered evacuations to create some disk space on the node. And since all nodes in a cluster tend to have same disk size, and azure-npm instances on each was generating logs with same velocity, most nodes starts to experience issues and triggered Pod evacuations in a very short timeframe at once. Some workload couldn't find healthy nodes and keep being impacted (keep being evacuated) for extended periods due to majority nodes in the cluster got hit by the issue.

**What you expected to happen:**  

* azure-npm to not to log onto node's root disk, or if needed there there should be some log rotation applied timely,
* azure-npm cpu/memory limits to be tuned to cover at least several hundreds of network policies in a cluster.

**How to reproduce it:**  
Create an AKS cluster, create 100 namespaces with this network policy and wait some weeks:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-ingress-from-same-namespace-and-infra
spec:
  podSelector: {}
  ingress:
  - from: 
    - podSelector: {}
  - from: 
    - namespaceSelector:
        matchLabels:
          infra: true
  egress:
  - {}
```

**Kubernetes Version:**  
v1.15.11

**Kernel (e.g. `uname -a`):**  
Linux aks-cpuworkers-34220213-vmss00000C 4.15.0-1082-azure #92~16.04.1-Ubuntu SMP Tue Apr 14 22:28:34 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Anything else we need to know?:**

azure-npm version: `mcr.microsoft.com/containernetworking/azure-npm:v1.1.4`

**Comment by csfmomo** [comment]
1. -azure-npm to not to log onto node's root disk, or if needed there there should be some log rotation applied timely,
    log rotation is already been added in npm v1.1.4 and has been fully tested. No more than 8 azure-npm.log will be generated. 
    Every log file will be limit in size of 5.5M. So no more than 50M log can npm generate in disk
    Could you check your npm log file? Could you provide how many azure-npm log file are you seeing and how large are they? 
    One suspicion could be if you are using npm version earlier than v1.1.4, it does have log flooding issue.
2. - azure-npm cpu/memory limits to be tuned to cover at least several hundreds of network policies in a cluster.
    We are going to increase it to 300 in the latest release. The previous version is 100.

One more notice, we are doing an iptable optimization for NPM. You won't see ~3K iptables rules in the coming NPM anymore.  Iptable rule number will be around the total network policy amount. This optimization will also help cpu.

**Comment by bergerx** [comment]
Thanks for the answer @csfmomo.

1. after your reply I immediately attempted to answer that the log rotation is not in place since I already provided the azure-npm version we have is v.1.1.4 as i was reporting this issue.

I decided to come back with evidence from our nodes and just found this to my surprise:
```
root@aks-cpuworkers02-32985644-vmss000019:/# ls -lhd /var/log/azure-npm*
-rw-r--r-- 1 root root 691K Jul  1 14:50 /var/log/azure-npm.log
-rw-r--r-- 1 root root 5.1M Jul  1 09:35 /var/log/azure-npm.log.1
-rw-r--r-- 1 root root  52G Jun 25 23:07 /var/log/azure-npm.log.2
root@aks-cpuworkers02-32985644-vmss000019:/# 
```

Seems like the v1.1.4 version is just deployed just after we experienced the issue and reporting the issue:
```
~ $ kubectl get ControllerRevision -l k8s-app=azure-npm --sort-by=revision
NAME                   CONTROLLER                 REVISION   AGE
azure-npm-858f59997c   daemonset.apps/azure-npm   1          103d
azure-npm-fc8f56bfd    daemonset.apps/azure-npm   4          89d
azure-npm-6fbf6f85d9   daemonset.apps/azure-npm   5          58d
azure-npm-56f46bf4c4   daemonset.apps/azure-npm   6          55d
azure-npm-b986c6dc6    daemonset.apps/azure-npm   7          48d
azure-npm-5797d5cdc6   daemonset.apps/azure-npm   8          26d
azure-npm-6bc9df44c5   daemonset.apps/azure-npm   9          5d16h
azure-npm-55c8f64c7c   daemonset.apps/azure-npm   10         5d15h
~ $ diff -u <(kubectl get ControllerRevision azure-npm-5797d5cdc6 -o json | jq .data.spec) <(kubectl get ControllerRevision azure-npm-6bc9df44c5 -o json | jq .data.spec)
--- /dev/fd/63	2020-07-01 15:39:45.745294429 +0100
+++ /dev/fd/62	2020-07-01 15:39:45.745594293 +0100
@@ -3,6 +3,7 @@
     "$patch": "replace",
     "metadata": {
       "annotations": {
+        "kubectl.kubernetes.io/restartedAt": "2020-06-25T22:52:32+01:00",
         "scheduler.alpha.kubernetes.io/critical-pod": ""
       },
       "creationTimestamp": null,
~ $ diff -u <(kubectl get ControllerRevision azure-npm-6bc9df44c5 -o json | jq .data.spec) <(kubectl get ControllerRevision azure-npm-55c8f64c7c -o json | jq .data.spec)
--- /dev/fd/63	2020-07-01 15:40:04.085739941 +0100
+++ /dev/fd/62	2020-07-01 15:40:04.086135882 +0100
@@ -48,7 +48,7 @@
               }
             }
           ],
-          "image": "mcr.microsoft.com/containernetworking/azure-npm:v1.1.2",
+          "image": "mcr.microsoft.com/containernetworking/azure-npm:v1.1.4",
           "imagePullPolicy": "IfNotPresent",
           "name": "azure-npm",
           "resources": {
~ $ 
```

For 2. This here is the current resource usage status (used the kubectl-status plugin) without any network policies. I'll try to share the stats after restoring the deleted NetworkPolicies:
![image](https://user-images.githubusercontent.com/45531/86259565-1475e700-bbb4-11ea-840c-d8b7d1da9c2f.png)

**Comment by bergerx** [comment]
Doesn't seem bad after enabling the NetworkPolicies again, here is resource utilisation in one of the azure-npm instances in our cluster:
![image](https://user-images.githubusercontent.com/45531/86264299-3a9e8580-bbba-11ea-9568-fdab18f3adfa.png)

I'll check the status of the written logs tomorrow.
